Proposed solution for dynamic RichTextBoxes attachment

mikhail-barg opened the pull request:
As discussed in https://github.com/NLog/NLog.Windows.Forms/issues/6 and https://github.com/NLog/NLog/issues/133

304NotModified commented:
It's not. Net 3.5 compatible? 

>ConcurrentDictionary' could not be found

mikhail-barg commented:
Also noticed that. I've thought that as NLog itself is now for 4.5, Forms should also be like that. Do you plan keeping 3.5 compatibility for NLog.Windows.Forms? If so I'll probably have to do some locking around regular Dictionary, but ConcurrentDictionary is surely preferred.

304NotModified commented:
NLog core is also 3.5+

304NotModified commented:
I would be nice to use the ConcurrentDictionary in net 4+

You need then multiple solutions (or maybe easier in vs2015?)

mikhail-barg commented:
OK then, I'll update to your latest changes and will see to add 3.5 compatibility.

304NotModified commented:
FYI: you can re-use the PR (just commit to your master branch) - that is also preferred.

304NotModified commented:
Just checked in short your code.  Why do we even need a (Concurrent) dictionary?

304NotModified commented:
*reopened this PR do you can add new commits on this branch

mikhail-barg commented:
The dictionary is needed to allow RTB controls to notify proper targets that they are available. Concurrent - so it would be thread-safe to do so, as targets may be added at run-time, right? Still I think the locking there would be of not much problem. Sorry that I haven't found time yet to re-write it for 3.5. Hope to do this tomorrow.

mikhail-barg commented:
I've realized that switching to a dictionary with locking is quite fast, so here's the update. 

Still I'm not sure why the ActiveFormNegativeTest1 is failing. I'm not convenient with XUnit, so I'm not sure how to debug it. As for me, it looks like it runs all tests in a same instance, so my glibal static dictionary is not cleared.

Could you please guide me with debugging this?

mikhail-barg commented:
I also thought on your question regarding the need for Dictionary. Is there a proper way to iterate through all the created targets? If yes, there would be no real need for Dictionary as additional global state.

mikhail-barg commented:
Ok, I swithed to enumerating through all targets in `LogManager.Configuration.AllTargets`, this way, there's no need for additional dictionary. Tests are working as well, except for the one I modified, as it seems that this is not en error anymore. Please have a look at this version.

304NotModified commented:
Thanks for this PR!

I have to admit, what is the real benefit? With this PR I can call

```c#
RichTextBoxTarget.RegisterTextBox(rtb);
```

But before it was
```c#
var target = LogManager.Configuration.FindTargetByName<RichTextBoxTarget>("name1");
if(target != null)
{
  target.TargetForm = from1; //or FormHelper.GetParentForm
  target.TargetRichTextBox = 

}
```

Yes, it's shorter, but also less transparent. 

I more in favor of:

```C#
//find all RichTextBoxes in the form an reinitialize them to all RichTextBoxTargets 
RichTextBoxTarget.ReInitializeAllTextboxes(form)

```

mikhail-barg commented:
Thanks for the interest in my proposal! 

Well, the main use-case that my proposal is aiming to support is to specify all targets in the configuration file, even for those textboxes that would appear on forms that would be opened later, and be able to do logging right away, knowing that when textboxes become available, they would start displaying logs, and no custom forms appear.

For that use-case your proposal with `RichTextBoxTarget.ReInitializeAllTextboxes(form)` does look better, as there's no need to specify each textbox. So I think that's the proper way. Do I get you right that we would need another method to notify targets when the form gets closed to remove links to controls (in analogy to my `unregister()` call), or how do you see this? Other solution would be to check `TargetTextBox.IsDisposed` in `Write`.

mikhail-barg commented:
I've merged the changes you did in the master, and updated code for using `RichTextBoxTarget.ReInitializeAllTextboxes(form)` and checking `TargetTextBox.IsDisposed`. 
Also restructurized code in `InitializeTarget()` so the flow is more clear. Should be fully backwards compatible.

304NotModified commented:
Thanks for the update!

I think `IsDisposed` is a good check. But I was also thinking:

Would it not be great if we use the [Form events](https://msdn.microsoft.com/en-us/library/system.windows.forms.form(v=vs.110).aspx), like `ControlAdded` and `FormClosed`? 

Because it would be nice if it works without calling `RichTextBoxTarget.ReInitializeAllTextboxes(form)`

mikhail-barg commented:
I would really like to get rid of need to notify logging system explicitly, but I don't think I understand how do you mean to use form events to make it work automatically. 

Events are attached to object instances, right? So we would need to attach our event to all forms somehow, and I don't know of a way to do this automatically (except making a Form descendant, which is no way better than adding `RichTextBoxTarget.ReInitializeAllTextboxes(this)` in a form constructor). And even if there's a way to attach to other form's events automatically, the `ControlAdded` event is actually a `Control.ControlAdded` event and it is raised when a _direct_ child is added to the control's `ControlCollection`. So to use this approach we would need to attach an event to every control, which clearly looks like an overkill.

It would ber nice to attach event to moment when a form is added to `Application.OpenForms` but I don't know how to do this (http://stackoverflow.com/questions/17850193/net-application-openforms-collection-changing, there's a way to do it like this: http://stackoverflow.com/a/1603721/376066 , but looks like a hack too).

In the end I think that requirement to call `RichTextBoxTarget.ReInitializeAllTextboxes(this)` from the form's constructor where you use logging to textboxes is the lesser evil. Or did I miss anything?

304NotModified commented:
>Events are attached to object instances, right? So we would need to attach our event to all forms somehow, and I don't know of a way to do this automatically

Well, we have the `Application.OpenForms` for that? Or are we then still to late?

> The ControlAdded event is actually a Control.ControlAdded event and it is raised when a direct child is added to the control's ControlCollection

I was thinking to hook the ControlAdded of the form, and when an textbox is added, we could check if that's the one we configured (check name) in the config.
Or do you think that won't work?

mikhail-barg commented:
> I was thinking to hook the ControlAdded of the form, and when an textbox is added, we could check if that's the one we configured (check name) in the config.
> Or do you think that won't work?

I think that this would not cover all the cases. Imagine we first add a groupbox to the form (at that time form's `ControlAdded` event is fired), and then add a richtextbox to the form (at that time groupbox's `ControlAdded` event is fired, but not form's, as far as I understand it).

> Well, we have the `Application.OpenForms` for that? Or are we then still to late?
I don't know how to attach any event handler to changes in `Application.OpenForms`. The only way I found is to monitor `WM_SHOWWINDOW` via `IMessageFilter` (see http://stackoverflow.com/a/1603721/376066), but this looks like an overkill. Or you are OK with this?

Or you mean to attach only to forms that are open at the moment target is initialized? Then yes, it would not work just as it does not work at the moment - if you have a field initializer for a logger for example, then the configuration would get initialized before any form is created.

304NotModified commented:
> I think that this would not cover all the cases.

You're right! 

I'm looking now in-depth to your code.

304NotModified commented on NLog.Windows.Forms.Tests/RichTextBoxTargetTests.cs:
```diff
@@ -442,7 +442,8 @@ public void ActiveFormNegativeTest1()
                     new LogFactory(config);
-                    Assert.True(false, "Expected exception.");
+                    //Actually this should work for now, no problem in having the target waiting for the richtextbox to appear, right?
+                    //Assert.True(false, "Expected exception.");
```
Well, backwards-compatiblity is really important, so if we should change unit tests, then something isn't backwards-compatible. We should fix that.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -82,6 +82,23 @@ static RichTextBoxTarget()
+            foreach (Target target in LogManager.Configuration.AllTargets)
+            {
+                RichTextBoxTarget rtbt = target as RichTextBoxTarget;
+                if (rtbt != null && rtbt.FormName == form.Name && rtbt.TargetRichTextBox == null)
```
please use English names for variables, like "textbox" etc (I know it was not before the 2.1.0).

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -207,6 +225,15 @@ public RichTextBoxTarget()
+        /// </remarks>
+        /// <docgen category='Form Options' order='10' />
+        [DefaultValue(true)]
+        public bool AllowCustomFormCreation { get; set; }
```
Proposal name: CreateFormIfNotFound

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -82,6 +82,23 @@ static RichTextBoxTarget()
+                {
+                    //can't use InitializeTarget here as the Application.OpenForms would not work from Form's constructor
+                    RichTextBox rtb = FormHelper.FindControl<RichTextBox>(rtbt.ControlName, form);
+                    if (rtb != null)
```
Please add some internallogger stuff

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -215,52 +242,66 @@ protected override void InitializeTarget()
             {
                 InternalLogger.Info("FormName not set, creating a new form");
-                FormName = "NLogForm" + Guid.NewGuid().ToString("N");
+                FormName = "NLogForm" + Guid.NewGuid().ToString("N");
```
`AllowCustomFormCreation ` must be `true` here for this?

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -215,52 +242,66 @@ protected override void InitializeTarget()
+            TargetForm = openFormByName;
+            CreatedForm = false;
+            TargetRichTextBox = FormHelper.FindControl<RichTextBox>(ControlName, TargetForm);
+            if (TargetRichTextBox == null)
```
this looks like code duplication. Please create a new private method

mikhail-barg replied:
Not sure what duplication you meant here, but I've extracted repeated code into `DoException` method.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -294,6 +335,12 @@ protected override void CloseTarget()
         {
+            RichTextBox rtbx = TargetRichTextBox;
+            if (rtbx == null || rtbx.IsDisposed)
+            {
```
add note to internal logger (debug)

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -82,6 +82,23 @@ static RichTextBoxTarget()
             DefaultRowColoringRules = rules.AsReadOnly();
         }
 
+        public static void ReInitializeAllTextboxes(Form form)
```
need for XML doc

304NotModified commented:
I added some notes. Small things I think.

Also, it would be perfect if I could call 
`ReInitializeAllTextboxes` without any arguments. 

It should loop over `Application.OpenForms` IMO. This need some changes in the current code, as `Application.OpenForms`  is `FormCollection` and not a single `Form` (and a foreach over ReInitializeAllTextboxes with 1 form is inefficient)

304NotModified commented:
Last, but not least, I think we should also create the dynamic approach into `FormControlTarget`? (we need maybe a new base class for this, to prevent code duplication). Do you agree?

mikhail-barg commented:
Thanks for the review! I'll try to work up all the noted things today.

> Last, but not least, I think we should also create the dynamic approach into `FormControlTarget`? (we need maybe a new base class for this, to prevent code duplication). Do you agree?

I haven't seen the `FormControlTarget` until you mentioned it, so I've no opinion right now, will hopefully look through it today too. What I currently see, is the `FindControlAndSendTheMessage` approach is not perfomance-efficient and whole idea behind doing this for RichTextBox target was to avoid that. 

So probably yes, it would be nice to refactor base functionality to base class.

mikhail-barg commented:
I've addressed all the things you've noted, except maybe for the duplicated code. Will check on the `FormControlTarget`

304NotModified commented:
Thanks for the update! :+1:

mikhail-barg commented:
I've looked through the FormControlTarget code and it looks that right now it has more differencies from the RichTextBoxTarget, than they have in common. For example to keep it backwards-compatible we will need to respect it's "check if control exists on every log message" behavior and make it to be default, which is quite different from RichTextBoxTarget's default "create new form if no control is found".

So in case we are to keep this behaviour, I only see it like writing it's own logics and then create some global method like `NlogWinForms.ReInitializeAllControls`, which would have separate handling for FormControl and RichTextBox targets.

mikhail-barg commented:
I will be not available for next 3 weeks, so please don't regard my silence as losing interest in this issue.

304NotModified commented:
OK thanks for the info. But I think it's almost done?

mikhail-barg commented:
Yes, I think it can be used as-is.

304NotModified commented:
I'm checking it out now. 

- ` CreateFormIfNotFound="false"` and no control name configured still creates a new form (did not expected that), maybe rename even `CreateFormIfNotFound` because it's in sync with the name.

304NotModified commented:
PS build will now fail if a public member has no XML docs

mikhail-barg commented:
Sorry for prolonged absence. Thanks for the review. I think I've addressed most of your comments in the code. Few notes:

> When it created the new form (`CreateFormIfNotFound = true`), then `RichTextBoxTarget.ReInitializeAllTextboxes` won't do anything (not expected)
Consider the following case. We have `CreateFormIfNotFound = true` and have logged something to the target, but the proper form was not available yet. So the custom form appeared, and it's textbox has few messages there. Now we have created the proper form and called `ReInitializeAllTextboxes()`, so the target can attach to the proper textbox. What should happen in this case?
I see the following possibilities:
1. nothing (as it's with my current code).
2. target attaches to textbox on the proper form (so next messages would get there), custom form remains open and unchanged.
3. target attaches to textbox on the proper form (so next messages would get there), custom form closed, messages logged there are lost.
4. target attaches to textbox on the proper form, moves previously logged messages to new textbox and closes the custom form.

I think only options 1 and 4 are not confusing to user, but option 4 requires storing all (or some in case of max line limit) messages until proper target is attached, so I'm not sure if it's a good idea. What do you think?

> log events before `RichTextBoxTarget.ReInitializeAllTextboxes` are lost. (nothing to do about it? We should document that clearly - in code and wiki)
This is symmetrical case but with `CreateFormIfNotFound = false`. If we stick with option 4 above, than this part would be solved as well.

mikhail-barg commented:
Would you please have another look at my PR, I'd really like it to get into release at some point.

304NotModified commented:
I will do. Sorry for the delay.

mikhail-barg commented:
No problem, just thought it got lost in the junk mail box )

304NotModified commented:
All the posts on github can be sometimes a bit overwhelming :p

But happy with the active community :)

304NotModified commented:
Will be upcoming week. Bit delayed by laptop (everything slow after Windows update).

304NotModified commented:
> . Now we have created the proper form and called ReInitializeAllTextboxes(), so the target can attach to the proper textbox. What should happen in this case?
> I see the following possibilities:
> 1. nothing (as it's with my current code).
> 2. target attaches to textbox on the proper form (so next messages would get there), custom form remains open and unchanged.
> 3. target attaches to textbox on the proper form (so next messages would get there), custom form closed, messages logged there are lost.
> 4. target attaches to textbox on the proper form, moves previously logged messages to new textbox and closes the custom form.
> 
> I think only options 1 and 4 are not confusing to user, but option 4 requires storing all (or some in case of max line limit) messages until proper target is attached, so I'm not sure if it's a good idea. What do you think?


I was expecting 2 or 3 ;) (really). It's called ReInitialize...not NOOP. And maybe the choice between 2 and 3 should be configurable. (Or the whole 1-4 options, but IMO 1 is bad and 4 is impossible)

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -202,62 +234,127 @@ public RichTextBoxTarget()
+        /// Called from constructor when error is detected. In case LogManager.ThrowExceptions is enabled, throws the exception, otherwise - logs the problem message
+        /// </summary>
+        /// <param name="message">exception/log text</param>
+        private static void HandleError(string message)
```
IMO a handleError with message parameters would be nice (`params object[]`), and no string concat in the callers. DO you agree? (not very important)

mikhail-barg replied:
Sure, no problem.

304NotModified commented:
I think we still have option 1?

mikhail-barg commented:
Right now - yes. I haven't done anything, waiting for your response.

So, considering your answer, I'll look towards implementing optional reinitialize strategies. 


> Is "accessory form" really the official/common terminology?

I just haven't invented a better name. Could you please suggest one?

mikhail-barg commented:
A bit aside from this thread, I was thinking of some other addition to the functionality of the target. We have a number of projects (standalone tools) that are working in the same environment and we'd like to have same logging behaviour for all of those projects. I.e. - same settings for network logging, same naming styles for logging to file, same list of email addresses for email target, etc. 

So we have a single nlog.config file that is shared between projects. And it works fine except for UI-based projects where we have RichTextBox targets which cannot be generalized unless we have same names for forms and controls in every project, which is not convenient at all.

I was thinking of some way to generalize the `FormName`  and `ControlName` elements of the RTB target, so the same config file could be used in different projects. What do you think?

304NotModified commented:
> So we have a single nlog.config file that is shared between projects. And it works fine except for UI-based projects where we have RichTextBox targets which cannot be generalized unless we have same names for forms and controls in every project, which is not convenient at all.

I don't fully understand this. Why this doesn't work?

mikhail-barg commented:
> I don't fully understand this. Why this doesn't work?

Imagine we have the target specified in the nlog.config:
```xml
<target name="textBox" xsi:type="RichTextBox" layout="${msg4rtb}"
            controlName="richTextBoxLog" formName="MainForm" autoScroll="true" maxLines="25">
      <row-coloring condition="level>=LogLevel.Error" fontColor="Red" style="Bold"/>
    </target>
```

To share it between many projects woe would need to name main form in every project as "MainForm". And if we have two forms in the project, only main form would have a common logging setup.
OR we would need to list all possible form names from all the projects in a shared config file, which is weird too. 

What I'm thinking of would be to have something like ` controlName="richTextBoxLog" formName="*" ` so every RTB control named "richTextBoxLog" on every form it could be fond in the project would get log messages as soon as it's available (thanks to newly introduced `Reinitialize` routine).
Or even ` controlName="*Log*" formName="*" `

Still this would mean having a single target for a number of textboxes (which should be not a huge change too). What do you think?

304NotModified commented:
Ah I understand 

What also is possible to make the formname layoutable, and use var, appsetting or some context layout renderer

mikhail-barg commented:
Yes, I also thought about using layout there, but it would not probably allow for capturing multiple controls at once, right?

304NotModified commented:
No that's true.  I'm doubting of * is really beter then a list.

mikhail-barg commented:
But with a list you would have to specify all possible form/control names explicitly, which is a bit backward - if you add a new form to some project, you'll need to update a shared config file to include new names, and all other projects became aware that you've done something completely unrelated to them.

304NotModified commented:
Good point. 

Well not sure if you want to include this in this PR. If so then:

- It should be backwardscompatible (so add another bool option I think)
-  I think a*b should be converted to regex. We need some really good code and tests on that. (we can include in NLog core (in a later phase)

mikhail-barg commented:
Well, we have managed to test the new code finally. The sad part is that we've lost the test setup we were using to investigate the initial issue (after all, it was more than two months ago) and we havent' had time to re-write it fully. Still we've run some simple tests, and it looks that everything works more or less as expected. 

Compared to previous version execution of some huge logging consumes not 100 MB of ram but merely 40 MB, which looks really nice. No speed delay was noticed, while we haven't run exact time measuring.

AsyncWrapper seem to work fine too.

What we have noticed though is some deadlock condition happening both with previous and curent code. The setup is like following:

config:
```xml
      <target xsi:type="Network"
        onOverflow="Discard"
        newLine="true"
        layout="${msg4Json}"
        encoding="utf-8"
        connectionCacheSize="5"
        keepConnection="false"
        address="tcp://logstash.our_local_host_name.com:9002"
              name="network"
      />
```

test program
```c#
        static void Main(string[] args)
        {
            Logger logger = LogManager.GetCurrentClassLogger();
            for (int i = 0; i < 100000; ++i)
            {
                logger.Trace("log " + i);
            };
            Console.WriteLine("Done");
            Console.Read();
        }
```

Call stacks for the old version of target:
```
Worker thread:
>	NLog.dll!NLog.Targets.NetworkTarget.Write.AnonymousMethod__a(System.Exception ex) + 0x76 bytes	
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend.AnonymousMethod__1a(System.Exception ex) + 0x74 bytes	
 	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.ProcessNextQueuedItem() + 0x10c bytes	
 	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.SocketOperationCompleted(object sender, System.Net.Sockets.SocketAsyncEventArgs e) + 0x1ad bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0x285 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0x9 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) + 0x57 bytes	
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure(System.Net.Sockets.SocketError socketError, int bytesTransferred, System.Net.Sockets.SocketFlags flags) + 0xa6 bytes	
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.CompletionPortCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* nativeOverlapped) + 0xa8 bytes	
 	mscorlib.dll!System.Threading._IOCompletionCallback.PerformIOCompletionCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* pOVERLAP) + 0x96 bytes	
 	[Native to Managed Transition]	

Main thread: 
> 	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.ProcessNextQueuedItem() + 0x64 bytes	
	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.DoSend(byte[] bytes, int offset, int length, NLog.Common.AsyncContinuation asyncContinuation) + 0x1f4 bytes	
 	NLog.dll!NLog.Internal.NetworkSenders.NetworkSender.Send(byte[] bytes, int offset, int length, NLog.Common.AsyncContinuation asyncContinuation) + 0x82 bytes	
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend.AnonymousMethod__1a(System.Exception ex) + 0x38c bytes	
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend(NLog.Internal.NetworkSenders.NetworkSender sender, byte[] buffer, NLog.Common.AsyncContinuation continuation) + 0x15b bytes	
 	NLog.dll!NLog.Targets.NetworkTarget.Write(NLog.Common.AsyncLogEventInfo logEvent) + 0x469 bytes	
 	NLog.dll!NLog.Targets.Target.WriteAsyncLogEvent(NLog.Common.AsyncLogEventInfo logEvent) + 0x1a4 bytes	
 	NLog.dll!NLog.LoggerImpl.WriteToTargetWithFilterChain(NLog.Internal.TargetWithFilterChain targetListHead, NLog.LogEventInfo logEvent, NLog.Common.AsyncContinuation onException) + 0x229 bytes	
 	NLog.dll!NLog.LoggerImpl.Write(System.Type loggerType, NLog.Internal.TargetWithFilterChain targets, NLog.LogEventInfo logEvent, NLog.LogFactory factory) + 0x23c bytes	
 	NLog.dll!NLog.Logger.WriteToTargets(NLog.LogLevel level, System.IFormatProvider formatProvider, string message) + 0xf9 bytes	
 	NLog.dll!NLog.Logger.Trace(string message) + 0x59 bytes	
 	NlogConsoleTest.exe!NlogConsoleTest.Program.Main(string[] args) Line 19 + 0x68 bytes	C#
```

call stacks for the new version of target:
```
Worker thread:
 	mscorlib.dll!System.Threading.Monitor.Enter(object obj, ref bool lockTaken) + 0x17 bytes	
>	NLog.dll!NLog.Targets.NetworkTarget.Write.AnonymousMethod__a(System.Exception ex) Line 321 + 0x1a bytes	C#
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend.AnonymousMethod__19(System.Exception ex) Line 485 + 0x10 bytes	C#
 	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.SocketOperationCompleted(object sender, System.Net.Sockets.SocketAsyncEventArgs e) Line 216 + 0x11 bytes	C#
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.OnCompleted(System.Net.Sockets.SocketAsyncEventArgs e) + 0x15 bytes	
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.ExecutionCallback(object ignored) + 0xf bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0xa7 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0x16 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) + 0x41 bytes	
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.FinishOperationSuccess(System.Net.Sockets.SocketError socketError, int bytesTransferred, System.Net.Sockets.SocketFlags flags) + 0x77e bytes	
 	System.dll!System.Net.Sockets.SocketAsyncEventArgs.CompletionPortCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* nativeOverlapped) + 0x37 bytes	
 	mscorlib.dll!System.Threading._IOCompletionCallback.PerformIOCompletionCallback(uint errorCode, uint numBytes, System.Threading.NativeOverlapped* pOVERLAP) + 0x6e bytes	
 	[Native to Managed Transition]	

Main thread:
 	mscorlib.dll!System.Threading.Monitor.Enter(object obj, ref bool lockTaken) + 0x17 bytes	
>	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.ProcessNextQueuedItem() Line 227 + 0x11 bytes	C#
 	NLog.dll!NLog.Internal.NetworkSenders.TcpNetworkSender.DoSend(byte[] bytes, int offset, int length, NLog.Common.AsyncContinuation asyncContinuation) Line 172 + 0x8 bytes	C#
 	NLog.dll!NLog.Internal.NetworkSenders.NetworkSender.Send(byte[] bytes, int offset, int length, NLog.Common.AsyncContinuation asyncContinuation) Line 115 + 0x16 bytes	C#
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend.AnonymousMethod__19(System.Exception ex) Line 511 + 0x37 bytes	C#
 	NLog.dll!NLog.Targets.NetworkTarget.ChunkedSend(NLog.Internal.NetworkSenders.NetworkSender sender, byte[] buffer, NLog.Common.AsyncContinuation continuation) Line 514 + 0x10 bytes	C#
 	NLog.dll!NLog.Targets.NetworkTarget.Write(NLog.Common.AsyncLogEventInfo logEvent) Line 316 + 0x47 bytes	C#
 	NLog.dll!NLog.Targets.Target.WriteAsyncLogEvent(NLog.Common.AsyncLogEventInfo logEvent) Line 204 + 0x34 bytes	C#
 	NLog.dll!NLog.LoggerImpl.WriteToTargetWithFilterChain(NLog.Internal.TargetWithFilterChain targetListHead, NLog.LogEventInfo logEvent, NLog.Common.AsyncContinuation onException) Line 196 + 0x28 bytes	C#
 	NLog.dll!NLog.LoggerImpl.Write(System.Type loggerType, NLog.Internal.TargetWithFilterChain targets, NLog.LogEventInfo logEvent, NLog.LogFactory factory) Line 99 + 0xe bytes	C#
 	NLog.dll!NLog.Logger.WriteToTargets(NLog.LogLevel level, System.IFormatProvider formatProvider, string message) Line 542 + 0x41 bytes	C#
 	NLog.dll!NLog.Logger.Trace(string message) Line 183 + 0x13 bytes	C#
 	TEST.exe!TEST.Program.Main(string[] args) Line 18 + 0x46 bytes	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	mscorlib.dll!System.AppDomain.ExecuteAssembly(string assemblyFile, System.Security.Policy.Evidence assemblySecurity, string[] args) + 0x6b bytes	
 	Microsoft.VisualStudio.HostingProcess.Utilities.dll!Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly() + 0x27 bytes	
 	mscorlib.dll!System.Threading.ThreadHelper.ThreadStart_Context(object state) + 0x6f bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0xa7 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) + 0x16 bytes	
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) + 0x41 bytes	
 	mscorlib.dll!System.Threading.ThreadHelper.ThreadStart() + 0x44 bytes	
 	[Native to Managed Transition]	
```

Considering that the issue happens both in old and new version, maybe it should be extracted as a separate issue?

mikhail-barg commented:
Please check my latest addition of MessageRetention strategy. After thinking more on the matter, I understood that you are right, and option 1 has no actual use. But I also think that keeping that "accessory" form open as in option 2 is also weird, considering that we could allow moving logged messages to new textbox.

Anyway, i think new solution is nice and flexible, thanks to your proposals! The only thing I'm not sure about right now is if we need to create new "accessory form" on a logging event if our current RichTextBox got disposed?

mikhail-barg commented:
Added automatic "accessory" form creation. Should be final now )

304NotModified commented:
I will review the PR this week

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -202,71 +235,231 @@ public RichTextBoxTarget()
+
+        /// <summary>
+        /// gets or sets the 
+        /// </summary>
```
Forgot something here ;)

mikhail-barg replied:
Fixed, thanks!

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -202,71 +235,231 @@ public RichTextBoxTarget()
+                    messageRetention = value;
+                    if (messageRetention == RichTextBoxTargetMessageRetentionStrategy.None)
+                    {
+                        if (messageQueue != null)
```
this `if` isn't really needed. Less is more

mikhail-barg replied:
Fixed.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -202,71 +235,231 @@ public RichTextBoxTarget()
+            //OnReattach?
+            switch (messageRetention)
+            {
+            case RichTextBoxTargetMessageRetentionStrategy.None:
```
please add some indenting.

mikhail-barg replied:
Sorry, this is the style I'm used to. Fixed.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -294,13 +496,60 @@ protected override void CloseTarget()
 
+            switch (messageRetention)
+            {
+            case RichTextBoxTargetMessageRetentionStrategy.None:
```
idem indenting

mikhail-barg replied:
Fixed

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -407,5 +657,41 @@ private void SendTheMessageToRichTextBox(string logMessage, RichTextBoxRowColori
+
+        private class MessageInfo
+        {
+            internal readonly string message;
```
I prefer (private) properties. IMO fields usage outside a class is not good pratice

mikhail-barg replied:
We have hearty discussion on this matter in our company, and I'm giving in to using fields ). Fixed.

304NotModified commented on NLog.Windows.Forms/RichTextBoxTarget.cs:
```diff
@@ -202,71 +235,231 @@ public RichTextBoxTarget()
+            get { return messageRetention; } 
+            set 
+            {
+                lock (messageQueueLock)
```
I think the scope of `lock` can be a lot smaller?

mikhail-barg replied:
I don't think so. I think all modifications to queue are to be synchronized. Anyway, the scope looks large because of brackets mostly )

304NotModified commented:
Looks really good! I have some last small things, see code comments. After that, I will merge and release it. 

We need also some docs on the wiki. Can you fix that? https://github.com/NLog/NLog.Windows.Forms/wiki/RichTextBoxTarget

- [ ] docs wiki

304NotModified commented:
If you like, some test cases could be added for the new code:

https://codecov.io/github/NLog/NLog.Windows.Forms/NLog.Windows.Forms/RichTextBoxTarget.cs?ref=3787d48f19ea14f538af495779ade2ff9b0acf73

mikhail-barg commented:
Handled most of the comments, except for the `lock` scope shortening. Thanks for the review!

> We need also some docs on the wiki. Can you fix that?

Sure, but should I wait for the release, or you'll be able to keep wiki changes unpublished before release?

> If you like, some test cases could be added for the new code:

Well, it's my responsibility, yes. But could you please guide me on installing Appveyor locally?

I've tried to do it myself and got a very moderate success. NuGet seem to have only v2.0 which is too modern for the version that you use. And without having it installed locally, I'll have to debug tests via commits, which is too unprofessional.

304NotModified commented:
> Well, it's my responsibility, yes. But could you please guide me on installing Appveyor locally?
> 
> I've tried to do it myself and got a very moderate success. NuGet seem to have only v2.0 which is too modern for the version that you use. And without having it installed locally, I'll have to debug tests via commits, which is too unprofessional.

No problem to help with AppVeyor. But I don't understand what you need right now. Do you like to run AppVeyor on your fork? (so locally = your fork, https://github.com/mikhail-barg/NLog.Windows.Forms)

Also you can run the unit tests just locally  (locally =your PC), no need for AppVeyor.

Least but not last, the code coverage is done by Codecov (and submitted by AppVeyor)

> Thanks for the review!
Thanks for the PR :+1:

304NotModified commented:
Wiki can be added live (just add "introduced in NLog.Windows.Forms 4.1")

Tesrs would be nice,  but I will merge this one already :)

mikhail-barg commented:
> But I don't understand what you need right now.

My bad. I actually meant XUnit, sorry.

> Also you can run the unit tests just locally (locally =your PC), no need for AppVeyor.

Yes, this is my case (locally == my PC). I recently upgraded to VS2015CE, so I should be able to install XUnit plugin into it, right?

304NotModified commented:
Doesn't xUnit work out of the box in VS? (I use Resharper - how can a developer live without it;) 

You can use the xUnit2 runner. It can run 1.x cases. 

Also, the unit tests can be upgraded to xUnit 2.1 You can then use https://www.nuget.org/packages/xunit.runner.visualstudio/

mikhail-barg commented:
Updated Wiki page.

304NotModified commented:
Thanks! I will release ťhe new version soon!
